patches/tModLoader/Terraria/ModLoader/NPCLoader.cs

```diff
@@ -35,12 +35,32 @@ public static class NPCLoader
 
 		private class HookList
 		{
-			public GlobalNPC[] arr = new GlobalNPC[0];
+			public int[] arr = new int[0];
 			public readonly MethodInfo method;
 
 			public HookList(MethodInfo method) {
 				this.method = method;
 			}
+
+			public IEnumerable<GlobalNPC> Enumerate(NPC npc) {
+				for (int i = 0; i < arr.Length; i++) {
+					var g = npc.globalNPCs[arr[i]];
```

Comment:
honestly it might be better for memory (and pretty similar or better for performance) to swap the loop indexing here. Have an `NPC` store the `globalNPCs` that are relevant for it, and then the hooklist have a `bool[]` for returning it in the iterator here.